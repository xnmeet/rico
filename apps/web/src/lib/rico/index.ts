import Rico from '@rico-core/parser';
import type { Document } from '@rico-core/parser';

export interface ParseResult {
  success: boolean;
  data?: string;
  error?: string;
}

export async function initRicoWasm() {
  try {
    await Rico.initialize();
  } catch (error) {
    console.error('Failed to initialize Rico WASM:', error);
    throw error;
  }
}
