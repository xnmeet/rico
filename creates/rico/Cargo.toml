[package]
name = "rico"
version = "0.1.6-alpha.0"
edition = "2021"
description = "A high-performance Apache Thrift IDL parser that converts Thrift IDL files to JSON AST"
license.workspace = true
documentation = "https://docs.rs/rico"
repository.workspace = true
readme = "README.md"
keywords = ["thrift", "parser", "ast", "idl", "json"]
categories = ["parsing", "development-tools"]
authors.workspace = true
exclude = ["tests/**/*", "snapshots/**/*"]
publish = true

[dependencies]
logos = "0.15.0"
serde = { workspace = true }
serde_json = { workspace = true }
miette = { workspace = true }
thiserror = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
