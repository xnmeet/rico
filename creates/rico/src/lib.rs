//! Rico is a high-performance Apache Thrift IDL parser and writer library.
//!
//! Rico provides functionality to parse Thrift IDL files into an AST (Abstract Syntax Tree)
//! and write the AST back to Thrift IDL format. It aims to be a complete solution for
//! working with Thrift IDL files in Rust.
//!
//! # Features
//!
//! - Fast and efficient parsing with detailed error reporting
//! - Complete Thrift IDL support including all standard constructs
//! - Bidirectional conversion between Thrift IDL and AST
//! - Preservation of comments and formatting
//! - Detailed source location tracking for debugging
//! - Support for annotations and custom metadata
//! - Clean and well-documented API
//!
//! # Modules
//!
//! - [`ast`]: Abstract Syntax Tree definitions and types. Provides the core data structures
//!   that represent Thrift IDL constructs in memory.
//!
//! - [`lexer`]: Tokenization of Thrift IDL input. Breaks down source text into a sequence
//!   of tokens for the parser to process.
//!
//! - [`parser`]: Parsing of tokens into AST. Implements recursive descent parsing with
//!   detailed error reporting and recovery.
//!
//! - [`writer`]: Converting AST back to Thrift IDL text. Handles proper formatting,
//!   indentation, and comment preservation.
//!
//! # Getting Started
//!
//! Rico can be used in two main ways:
//!
//! 1. Parsing Thrift IDL into AST for analysis or transformation
//! 2. Writing AST back to Thrift IDL text
//!
//! ## Parsing Example
//!
//! ```rust
//! use rico::Parser;
//!
//! fn main() {
//!     let input = r#"
//!         namespace rs demo
//!
//!         struct User {
//!             1: string name
//!             2: i32 age
//!         }
//!     "#;
//!
//!     let mut parser = Parser::new(input);
//!     match parser.parse() {
//!         Ok(ast) => println!("{}", serde_json::to_string_pretty(&ast).unwrap()),
//!         Err(e) => eprintln!("Error: {}", e),
//!     }
//! }
//! ```
//!
//! ## Writing Example
//!
//! ```rust
//! use rico::{Parser, Writer};
//!
//! fn main() {
//!     let input = r#"
//!         struct User {
//!             1: string name
//!             2: i32 age
//!         }
//!     "#;
//!
//!     // Parse the input into an AST
//!     let mut parser = Parser::new(input);
//!     if let Ok(ast) = parser.parse() {
//!         // Create a writer and convert AST back to Thrift IDL
//!         let mut writer = Writer::new();
//!         let output = writer.write(&ast);
//!         println!("{}", output);
//!     }
//! }
//! ```
//!
//! # Supported Thrift Features
//!
//! ## Types
//! - Base types (bool, byte, i16, i32, i64, double, string, binary)
//! - Container types (list, set, map)
//! - User-defined types (struct, union, exception, enum)
//!
//! ## Definitions
//! - Constants with complex values (numbers, strings, lists, maps)
//! - Typedefs for type aliases and documentation
//! - Enums with optional values and annotations
//! - Structs with required/optional fields and defaults
//! - Unions for variant types with field IDs
//! - Exceptions for error handling with inheritance
//! - Services with inheritance and function modifiers
//!
//! ## Other Features
//! - Namespaces for different target languages
//! - Include statements for modular IDL organization
//! - Field IDs and requiredness specifiers
//! - Default values for fields
//! - Oneway function modifiers
//! - Throws clauses for error handling
//! - Single and multi-line comments
//! - Annotations for metadata and custom attributes
//!
//! # Error Handling
//!
//! The library provides detailed error reporting for both parsing and writing operations.
//! Errors include:
//! - Source location information (line and column)
//! - Descriptive error messages
//! - Expected vs actual token information
//! - Context about the construct being processed
//! - Suggestions for fixing common issues
//!
//! # Best Practices
//!
//! 1. Always check for errors when parsing
//! 2. Use the writer to maintain consistent formatting
//! 3. Preserve comments and annotations when modifying AST
//! 4. Handle all error cases appropriately
//! 5. Validate AST modifications before writing

pub mod ast;
pub mod lexer;
pub mod parser;
pub mod writer;

pub use ast::*;
pub use parser::Parser;
pub use writer::Writer;
