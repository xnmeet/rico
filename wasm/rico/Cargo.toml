[package]
name = "rico-wasm"
version = "0.0.1"
edition = "2021"
description = "WebAssembly bindings for Rico - A high-performance Apache Thrift IDL parser and writer library."
license.workspace = true
documentation = "https://docs.rs/rico"
repository.workspace = true
readme = "README.md"
keywords = ["thrift", "parser", "ast", "idl", "json"]
categories = ["parsing", "development-tools"]
authors.workspace = true
publish = false

[lib]
crate-type = ["cdylib", "rlib"]

[features]
default = ["console_error_panic_hook"]

[dependencies]
rico = { workspace = true }
wasm-bindgen = "0.2"
serde-wasm-bindgen = "0.6"
serde = { workspace = true }
miette = { workspace = true }
# The `console_error_panic_hook` crate provides better debugging of panics by
# logging them with `console.error`. This is great for development, but requires
# all the `std::fmt` and `std::panicking` infrastructure, so isn't great for
# code size when deploying.
console_error_panic_hook = { version = "0.1.7", optional = true }


[profile.release]
# Tell `rustc` to optimize for small code size.
opt-level = "s"
