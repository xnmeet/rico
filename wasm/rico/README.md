# @rico-core/parser

WebAssembly bindings for Rico - A high-performance Apache Thrift IDL parser and writer library.

## Installation

```bash
npm install @rico-core/parser
```

## Usage

```typescript
import Rico, { Document, Struct } from '@rico-core/parser';

async function main() {
  // Initialize Rico (required before any operations)
  await Rico.initialize();

  const input = `
    namespace rs demo

    struct User {
      1: string name
      2: i32 age
    }
  `;

  try {
    // Parse Thrift IDL to AST with full type information
    const ast: Document = await Rico.parse(input, false);

    // Type-safe access to AST properties
    ast.members.forEach(member => {
      if (member.kind === 'StructDefinition') {
        const struct = member as Struct;
        console.log('Found struct:', struct.name.value);
        struct.members.forEach(field => {
          console.log(
            ` - ${field.name.value}: ${field.fieldType.value} (ID: ${field.fieldID?.value})`
          );
        });
      }
    });

    // Write AST back to Thrift IDL
    const output = Rico.write(ast);
    console.log('Generated Thrift IDL:', output);
  } catch (error) {
    console.error('Error:', error);
  }
}

main().catch(console.error);
```

## API

### `Rico.initialize(): Promise<void>`

Initializes the WebAssembly module. Must be called before using any other methods.

### `Rico.parse(input: string): Promise<Document>`

Parses a Thrift IDL string and returns the AST (Abstract Syntax Tree) with full type information.

### `Rico.write(ast: Document): string`

Converts an AST back to Thrift IDL format.

## Type System

Rico provides a comprehensive type system that exactly matches the Rust AST definitions. All types are automatically generated from the Rust source code to ensure perfect alignment.

### Core Types

- `Document`: The root AST node
- `DocumentMember`: Union type of all possible top-level definitions
- `BaseNode`: Common properties for all AST nodes
- `NodeType`: Enum of all possible node types

### Definition Types

- `Namespace`: Namespace declarations
- `Include`: Include statements
- `Const`: Constant definitions
- `Typedef`: Type aliases
- `Enum`: Enum definitions
- `Struct`: Struct definitions
- `Union`: Union definitions
- `Exception`: Exception definitions
- `Service`: Service definitions

### Field Types

- `Field`: Field definitions in structs/unions/exceptions
- `FieldType`: Union type of all possible field types
  - `CommonType`: Basic types (string, i32, etc.)
  - `FieldSetType`: Sets
  - `FieldListType`: Lists
  - `MapType`: Map types with key and value types

### Value Types

- `FieldValue`: Union type of all possible values
  - `ConstValue`: Basic constant values
  - `ConstList`: List literals
  - `ConstMap`: Map literals

### Location Information

All nodes include source location information:

```typescript
interface LOC {
  start: Span;
  end: Span;
}

interface Span {
  line: number;
  column: number;
  index: number;
}
```

### Type Safety Example

```typescript
import { Document, Struct, Field, FieldType } from '@rico-core/parser';

function processStruct(struct: Struct) {
  // All fields are properly typed
  struct.members.forEach((field: Field) => {
    // Type-safe field access
    const fieldName = field.name.value;
    const fieldType = field.fieldType;
    const isRequired = field.requiredType === 'required';

    // Type-safe pattern matching on field types
    if (fieldType.kind === 'CollectionType') {
      console.log(
        `${fieldName} is a collection of ${fieldType.valueType.value}`
      );
    } else if (fieldType.kind === 'MapType') {
      console.log(
        `${fieldName} is a map from ${fieldType.keyType.value} to ${fieldType.valueType.value}`
      );
    }
  });
}
```

## Development

1. Build the WebAssembly module:

```bash
npm run build
```

2. Run tests:

```bash
npm test
```

## License

MIT
