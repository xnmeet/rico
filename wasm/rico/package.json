{
  "name": "@rico-core/parser",
  "version": "0.0.4",
  "description": "WebAssembly bindings for Rico - A high-performance Apache Thrift IDL parser",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "collaborators": [
    "xxnmeet@gmail.com"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/xnmeet/rico"
  },
  "scripts": {
    "clean": "rm -rf dist src/wasm",
    "pack": "wasm-pack build ../rico --target web --out-dir ./src/wasm",
    "move-wasm": "mkdir -p dist/wasm && cp src/wasm/*.{ts,js,wasm} dist/wasm/",
    "build": "npm run clean && npm run pack && tsc && npm run move-wasm"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "thrift",
    "idl",
    "parser",
    "wasm",
    "webassembly"
  ],
  "author": "xxnmeet@gmail.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@types/node": "^20.0.0",
    "typescript": "^5.0.0",
    "wasm-pack": "^0.13.0"
  }
}
