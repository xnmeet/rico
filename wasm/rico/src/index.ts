import init, { Parser, Writer } from './wasm/rico_wasm';
import type { Document, ParseError } from './types';

export class RicoError extends Error {
  constructor(public details: ParseError) {
    const message = [
      `${details.message}${
        details.location &&
        `(${details.location.line}:${details.location.column})`
      }`,
      details.help && `Help: ${details.help}`
    ]
      .filter(Boolean)
      .join('\n');

    super(message);
    this.name = 'RicoError';
    // Prevent V8 from collecting stack trace
    Error.captureStackTrace(this, RicoError);
  }
}

export class Rico {
  private static initialized = false;

  static async initialize(): Promise<void> {
    if (!Rico.initialized) {
      await init();
      Rico.initialized = true;
    }
  }

  static parse(input: string): Promise<string>;
  static parse(input: string, toString: false): Promise<Document>;
  static parse(input: string, toString: true): Promise<string>;
  static parse(
    input: string,
    toString: boolean = true
  ): Promise<Document | string> {
    if (!Rico.initialized) {
      throw new Error('Rico is not initialized. Call Rico.initialize() first.');
    }
    const parser = new Parser(input);
    try {
      const result = parser.parse();
      return Promise.resolve(!toString ? JSON.parse(result) : result);
    } catch (error) {
      try {
        const parserError = JSON.parse(error as string);
        if (parserError && 'kind' in parserError) {
          throw new RicoError(error as ParseError);
        }
      } finally {
        throw error;
      }
    }
  }

  static write(ast: Document): string {
    if (!Rico.initialized) {
      throw new Error('Rico is not initialized. Call Rico.initialize() first.');
    }
    const writer = new Writer();
    try {
      return writer.write(JSON.stringify(ast));
    } catch (error) {
      try {
        const parserError = JSON.parse(error as string);
        if (parserError && 'kind' in parserError) {
          throw new RicoError(error as ParseError);
        }
      } finally {
        throw error;
      }
    }
  }
}

export * from './types';
export default Rico;
