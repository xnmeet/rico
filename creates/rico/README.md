# Thrift Parser

A high-performance Apache Thrift IDL parser written in Rust that converts Thrift IDL files to JSON AST.

## Features

- 🚀 Fast and efficient parsing
- 🎯 Complete Thrift IDL support
- 🔄 JSON AST output
- 📝 Comment preservation
- 🎨 Detailed source location tracking
- ⚡ Parallel processing support
- 📊 Built-in benchmarking

## Installation

```bash
[dependencies]
rico_parser = "0.1.0"
```

## Usage

### Basic Parsing

```rust
use rico::Parser;

fn main() {
    let input = r#"
        namespace rs demo

        struct User {
            1: string name
            2: i32 age
        }
    "#;

    let mut parser = Parser::new(input);
    match parser.parse() {
        Ok(ast) => println!("{}", serde_json::to_string_pretty(&ast).unwrap()),
        Err(e) => eprintln!("Error: {}", e),
    }
}
```

### Command Line Tools

#### Scan Multiple Files

The project includes a CLI tool for processing multiple Thrift files:

```bash
cargo run -p thrift-scan -- --path ./thrift/files --output ./json/output
```

#### Benchmark Parser Performance

Run benchmarks on Thrift files:

```bash
cargo run -p thrift-benchmark
```

## Project Structure

- `lib/`: Core parser library
- `scan/`: CLI tool for batch processing
- `benchmark/`: Performance benchmarking tool

## Supported Thrift Features

- Base types (i32, i64, string, etc.)
- Collections (list, set, map)
- Structs and Exceptions
- Services and Functions
- Enums
- Constants
- Typedefs
- Namespaces
- Includes
- Comments and Annotations

## Development

### Building

```bash
cargo build --workspace
```

### Running Tests

```bash
cargo test --workspace
```

### Code Structure

- Lexer: Tokenizes input using Logos
- Parser: Recursive descent parser
- AST: Strongly typed syntax tree
- Location Tracking: Preserves source positions

## Contributing

1. Fork the repository
2. Create your feature branch (`git checkout -b feature/amazing-feature`)
3. Commit your changes (`git commit -m 'Add some amazing feature'`)
4. Push to the branch (`git push origin feature/amazing-feature`)
5. Open a Pull Request

## License

This project is licensed under the MIT License - see the LICENSE file for details.

## Acknowledgments

- [Logos](https://github.com/maciejhirsz/logos) for lexer generation
- [Serde](https://github.com/serde-rs/serde) for JSON serialization
