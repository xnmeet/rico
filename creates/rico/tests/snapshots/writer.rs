use insta::assert_snapshot;
use miette::NamedSource;
use rico::{Parser, Writer};
use std::fs;
use std::path::{Path, PathBuf};

#[test]
pub fn test_writer_snapshots() {
    let tests_dir = get_tests_dir();
    let thrift_files = collect_thrift_files(&tests_dir);

    let snapshot_dir = std::env::current_dir().unwrap().join("snapshots/writer");

    fs::create_dir_all(&snapshot_dir).unwrap();

    for path in thrift_files {
        let content = fs::read_to_string(&path).unwrap();
        let file_name = path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or_default();

        let mut parser = Parser::new(&content);
        let result = parser
            .parse()
            .map_err(|err| {
                println!(
                    "{:?}",
                    miette::Error::new(err).with_source_code(NamedSource::new(
                        path.display().to_string(),
                        content.clone(),
                    ))
                );
            })
            .unwrap();

        let mut writer = Writer::new();
        let thrift_code = writer.write(&result);

        insta::with_settings!({
            description => format!("Testing {} thrift file", file_name),
            omit_expression => true,
            snapshot_path => snapshot_dir.clone(),
        }, {
            assert_snapshot!(file_name, thrift_code);
        });
    }
}

fn get_tests_dir() -> PathBuf {
    let current_dir = std::env::current_dir().unwrap();
    if current_dir.join("tests").exists() {
        current_dir.join("tests/fixtures")
    } else {
        current_dir.parent().unwrap().join("tests/fixtures")
    }
}

fn collect_thrift_files(dir: &Path) -> Vec<PathBuf> {
    let mut files = Vec::new();
    if dir.is_dir() {
        for entry in fs::read_dir(dir).unwrap() {
            let entry = entry.unwrap();
            let path = entry.path();
            if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("thrift") {
                files.push(path);
            }
        }
    }
    files.sort();
    files
}
