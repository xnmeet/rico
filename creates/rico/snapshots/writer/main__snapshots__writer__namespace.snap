---
source: creates/rico/tests/snapshots/writer.rs
description: Testing namespace thrift file
snapshot_kind: text
---
namespace py main_service

namespace go other_service

namespace go a.b.c
